channeldb/mp_payment.go

```diff
@@ -359,6 +348,86 @@ func (m *MPPayment) UpdateState() error {
 	return m.updateState()
 }
 
+// NeedWaitAttempts decides whether we need to hold creating more HTLC attempts
+// and wait for the results of the payment's inflight HTLCs. Return an error if
+// the payment is in an unexpected state.
+func (m *MPPayment) NeedWaitAttempts() (bool, error) {
+	// Check when the remainingAmt is not zero, which means we have more
+	// money to be sent.
+	if m.State.RemainingAmt != 0 {
+		switch m.Status {
+		// If the payment is newly created, no need to wait for HTLC
+		// results.
+		case StatusInitiated:
+			return false, nil
+
+		// If we have inflight HTLCs, we'll check if we have terminal
+		// states to decide if we need to wait.
+		case StatusInFlight:
+			// We still have money to send, yet one of the HTLCs is
+			// settled, return an error as the peer is violating
+			// the protocol.
+			if m.State.HasSettledHTLC {
+				return false, fmt.Errorf("%w: settled HTLC "+
+					"but still have remaining amount %v",
+					ErrPaymentInternal,
+					m.State.RemainingAmt)
+			}
+
+			// Otherwise we don't need to wait for inflight HTLCs
+			// since we still have money to be sent.
+			return false, nil
+
+		// We need to send more money, yet the payment is already
+		// succeeded. Return an error in this case as the peer is
+		// violating the protocol.
+		case StatusSucceeded:
+			return false, fmt.Errorf("%w: payment already "+
+				"succeeded but still have remaining amount %v",
+				ErrPaymentInternal, m.State.RemainingAmt)
+
+		// The payment is failed and we have no inflight HTLCs, no need
+		// to wait.
+		case StatusFailed:
+			return false, nil
+
+		// Unknown payment status.
+		default:
+			return false, fmt.Errorf("%w: %s",
+				ErrUnknownPaymentStatus, m.Status)
+		}
+	}
+
+	// Now we determine whether we need to wait when the remainingAmt is
+	// already zero.
+	switch m.Status {
+	// When the payment is newly created, yet the payment has no remaining
+	// amount, return an error.
+	case StatusInitiated:
+		return false, fmt.Errorf("%w: %v", ErrPaymentInternal, m.Status)
+
+	// If the payment is inflight, we must wait.
+	//
+	// NOTE: an edge case is when all HTLCs are failed while the payment is
+	// not failed we'd still be in this inflight state. However, since the
+	// remainingAmt is zero here, it means we cannot be in that state as
+	// otherwise the remainingAmt would not be zero.
```

Comment:
Q: should we not return an error, because we should never end in this state, or at least a warning that we ended in an unexpected state ?

Reply:
> Does that edge case exist? Looking at decidePaymentStatus, inflight = true is not reached when all HTLCs are failed (when we check each HTLC in the for loop and they are all failed, we don't set inflight=true).

This would be the case `htlcFailed` in `decidePaymentStatus`, which returns us `StatusInFlight`. This is a hidden state IMO. A previous design [explicitly expressed this state](https://github.com/yyforyongyu/lnd/commit/c29cffcb5f7697df606a7917d829770090de8e01#diff-81823a7e88e5712fc053292cdbff006c9c58796c27c16cf2af9d41e64241d49eR20), and ditched because there were too many states, another tradeoff we made.

> Q: should we not return an error, because we should never end in this state, or at least a warning that we ended in an unexpected state ?

Unfortunately the current design cannot distinguish that state. This can only happen if our current view of the payment thinks the HTLCs are still ongoing, and they are failed in other goroutines, which means we won't have the updated view until the next iteration. For this case it's safe because we'll wait for the HTLC attempts. Once they are marked as failed in db, the `remainingAmt` would then become non-zero, and we can continue the lifecycle.